### ISSUE TITLE
Player not playing when offline 

### ISSUE THREAD

**Issue by Rohan-Ramesh-Jadhav** [shaka-project/shaka-player#4120]
Hello, I am trying to play the stored content by turning off my WiFi and checking offline play. The Player did not play and gives results in dev tool console as shown in the screenshot. Can anyone redirect me where I am falling short or making mistake. Below is the screen short and code For the issue.
Thank you.
`window.player = new shaka.Player(video);`
`window.storage.list().then((res)=>{`
`let offlineData = res[0];`
`window.player.load(offlineData['offlineUri']).then((res)=>{console.log(res)}).catch((err)=>{console.log(err)});`
`});`
Where 'video' is the HTML video tag.
![offlineError](https://user-images.githubusercontent.com/58617785/162883082-6ae10495-195c-43c3-b799-8524c53a6b41.png)

**Comment by Rohan-Ramesh-Jadhav** [comment]
I tried to attach the player to the same video tag, but I couldn't do it, so after every song I destroy the whole video element and attach a new video tag. This is how destroy the tag and attach new tag
       ` $('#jVideoPlayer').remove();`
       `let videoTag = document.createElement('video');`
        `videoTag.setAttribute('id', 'jVideoPlayer');`
        `videoTag.setAttribute('style', 'width:250px;height:250px;');`
        `videoTag.setAttribute('controlsList', 'nodownload noplaybackrate');`
        `videoTag.setAttribute('autoplay','true');`
        `main_div.appendChild(videoTag);`
Even when there is no internet I do the same.
Is that I am doing it wrong or is it fine?

**Comment by theodab** [comment]
Why not just call `player.unload()` after a song is over, instead of remaking the player entirely?

**Comment by Rohan-Ramesh-Jadhav** [comment]
Thank you @theodab for the suggession on player.unload() it worked like charm for me on unloading the player. But offline play issue persist as same I am getting the same result as it is in the screenshot above. If you could you help me out on this then it would be great help brother.
Thank you

**Comment by Rohan-Ramesh-Jadhav** [comment]
I think there will be someone who may help me out of this please.

**Comment by theodab** [comment]
Are you still making a new player instance when you run offline? If you really need to for some reason, you should call `player.detach()` on the old player instance (and wait for the promise to resolve) before making a new one.
I can't give much more advice just going by code snippets, though.

**Comment by Rohan-Ramesh-Jadhav** [comment]
@joeyparrish thank you for your response, I will try the same and will come back to you.

**Comment by Rohan-Ramesh-Jadhav** [comment]
Hello @joeyparrish,
I used the player.detach() as suggested and it loads the data also, but player doesn't play even though i tap play button of the video tag. Below is what I am doing for offline play
`//detach code`
`      let video = document.getElementById('jVideoPlayer');`
`      window.player.detach().then((res)=>{`
`        window.player = new shaka.Player(video);`  
`        window.storage.list().then((res)=>{`
`          let offlineData = res[0];`
`          window.player.load(offlineData.offlineUri)`
`          .then(() => {`
`            console.log('player Play');`
`          })`
`          .catch((error) =>`
`          console.error("Error code", error.code, "object", error)`
`          );`
`      });`
`      }).catch((err)=>{`
`        console.log(err);`
`      });`

By the above code, it gives the console log as player play, but the video doesn't play. Could you help me out with this?
Thank you.

**Comment by joeyparrish** [comment]
Your code is very hard to read formatted in this way.  Here's a markdown tip: Use ` ```js ` on its own line, then all your code, then another ` ``` ` on its own line.  That will format the entire block as code and syntax highlight it for JavaScript.  I've edited your last post in this way, and also reindented the code for clarity.

The problem I see in your code is that you're detaching from the video element, but then creating a new player.  You don't need to do that.  Instead, you should await player.attach(video).

But from your code snippet, it is unclear why you need to detach/attach at all.  You should be able to transition from one song to another like this:

```js
  const onlineSongUrl = '...';  // Fill this in
  await player.load(onlineSongUrl);

  // later, when you want to play a different song
  const storedOffline = await window.storage.list();
  const offlineSongUrl = storedOffline[0].offlineUri;
  await player.load(offlineSongUrl);
```

Here are the main lifecycle methods and their purposes:
  - new Player(mediaElement): Creates a player and takes control of a media element.
  - new Player(): Creates a player without control of a media element.
  - attach(mediaElement): Take control of a media element.
  - load(): Load content.  Implicitly unloads any content already loaded.  Must have control of a media element.
  - unload(): Unload content.  Does not give up control of the media element.
  - detach(): Give up control of the media element.  Implicitly unloads any content already loaded.
  - destroy(): Destroy the player.  Implicitly unloads content and detaches from the media element.

You don't need more than one Player instance if you only play one thing at a time.  You don't need more than one media element if you only play one thing at a time.  If you only play one thing at a time, just call load() for each thing you want to play.  detach() and attach() are only needed if you want to time-share the media element with some other player.

Does this help?

**Comment by Rohan-Ramesh-Jadhav** [comment]
Thank you @joeyparrish above instructions helped me in one way. I understood that every time I must not create a player and attach the player. and I changed the code in my project to the same. It worked like charm. But still, music didn't play when I turned off the internet connection to my PC. I will do one thing I will place my code to store the data and play offline. Could you able to tell me where I am going wrong? I am new to this and trying to learn how to use the player for DRM. Please do help me out of this.
Thank you for sparing your valuable time for me
```js
// code how I am storing the data when online
window.storage = new shaka.offline.Storage(window.player);
window.storage.configure({
    offline: {
      usePersistentLicense: false
    }
});
const sMetadata = {
   'title':trackName
};
window.storage.store(manifestUri,sMetadata);

// code of offline play
function offlinePlay()
{
//offline test
  let offlineCheck = setInterval(()=>{
    if(!navigator.onLine)
    {
      window.clearInterval(offlineCheck);
//detach the video tag
      window.storage.list().then((res)=>{
        let offlineMainfest = res[0].offlineUri;
        window.player.load(offlineMainfest).then(()=>{
          console.log(window.player.getStats());
        }).catch((err)=>{
          console.log(err);
        });
      }).catch((err)=>{
        console.log(err);
      });
    }
    console.log('offline');
  },1000);
// offline test ends
}
offlinePlay();
```
Sorry for the inconvineance because of code wasn't indented last time and thankyou for the suggestion.

**Comment by joeyparrish** [comment]
I see you're calling `window.storage.store(manifestUri,sMetadata)` without waiting for the async process to complete.  That could be an issue.

I also see that you're switching from streaming to offline playback based on a timer.  This is unnecessary, and will create small playback issues when the switch occurs, such as the user losing their place in the content.  Instead, I recommend you store the content, wait for this to complete, then immediately begin playing the stored version.

You can play content from a database even when you're online.  The user doesn't spend twice the bandwidth (once to store, once more to stream it again), and they don't lose their place in the content when they transition from WiFi to offline or back again.

Does this help?
